google-cloud-bigquery/src/main/scala/akka/stream/alpakka/googlecloud/bigquery/scaladsl/BigQueryQueries.scala

```diff
@@ -0,0 +1,157 @@
+import akka.stream.RestartSettings
+import akka.stream.alpakka.googlecloud.bigquery.impl.http.BigQueryHttp
+import akka.stream.alpakka.googlecloud.bigquery.model.QueryJsonProtocol.{QueryRequest, QueryResponse}
+import akka.stream.alpakka.googlecloud.bigquery.{BigQueryAttributes, BigQueryEndpoints, BigQueryException}
+import akka.stream.scaladsl.{Keep, RestartSource, Sink, Source}
+
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+import scala.util.{Failure, Success}
+
+private[scaladsl] trait BigQueryQueries { this: BigQueryRest =>
+
+  /**
+   * Runs a BigQuery SQL query.
+   * @see [[https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs/query BigQuery reference]]
+   *
+   * @param query a query string, following the BigQuery query syntax, of the query to execute
+   * @param dryRun if set to `true` BigQuery doesn't run the job and instead returns statistics about the job such as how many bytes would be processed
+   * @param useLegacySql specifies whether to use BigQuery's legacy SQL dialect for this query
+   * @tparam Out the data model of the query results
+   * @return a [[akka.stream.scaladsl.Source]] that emits an [[Out]] for each row of the results and materializes a [[scala.concurrent.Future]] containing the [[akka.stream.alpakka.googlecloud.bigquery.model.QueryJsonProtocol.QueryResponse]]
+   */
+  def query[Out](query: String, dryRun: Boolean = false, useLegacySql: Boolean = true)(
+      implicit um: FromEntityUnmarshaller[QueryResponse[Out]]
+  ): Source[Out, Future[QueryResponse[Out]]] = {
+    val request = QueryRequest(query, None, None, None, Some(dryRun), Some(useLegacySql), None)
+    this.query(request)
+  }
+
+  /**
+   * Runs a BigQuery SQL query.
+   * @see [[https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs/query BigQuery reference]]
+   *
+   * @param query the [[akka.stream.alpakka.googlecloud.bigquery.model.QueryJsonProtocol.QueryRequest]]
+   * @tparam Out the data model of the query results
+   * @return a [[akka.stream.scaladsl.Source]] that emits an [[Out]] for each row of the results and materializes a [[scala.concurrent.Future]] containing the [[akka.stream.alpakka.googlecloud.bigquery.model.QueryJsonProtocol.QueryResponse]]
+   */
+  def query[Out](query: QueryRequest)(
+      implicit um: FromEntityUnmarshaller[QueryResponse[Out]]
+  ): Source[Out, Future[QueryResponse[Out]]] =
+    Source
+      .fromMaterializer { (mat, attr) =>
+        import BigQueryException._
+        import SprayJsonSupport._
+        import mat.executionContext
+        implicit val system = mat.system
+        implicit val settings = BigQueryAttributes.resolveSettings(attr, mat)
+
+        Source.lazyFutureSource { () =>
+          for {
+            entity <- Marshal(query).to[RequestEntity]
+            initialRequest = HttpRequest(POST, BigQueryEndpoints.queries(settings.projectId), entity = entity)
+            response <- BigQueryHttp().retryRequestWithOAuth(initialRequest)
```

Comment:
The source should fail right? Is this not the desired behavior?

Reply:
It should fail, but I think because you're using the for comprehension to resolve this intermediate `Future` any failure will just get swallowed and the end result would be the `Source` getting stuck.

I reproduced that behaviour with this sample.

```
        def fail: Future[String] = Future(throw new Exception("error"))

        val source: Source[String, Future[NotUsed]] = Source.lazyFutureSource { () =>
          for {
            r1 <- Future.successful("f1")
            r2 <- fail
          } yield Source(List(r1, r2))
        }

        val f = source.runWith(Sink.collection)

        val strs = Await.result(f, 5.seconds)
```

I could be missing something though..

Reply:
Really? I just ran your code snippet and it throws the exception.

Reply:
That's strange. I added the snippet as a test in `BigQueryQueriesSpec` and [pushed a branch](https://github.com/seglo/alpakka/tree/seglo/wip-new-bigquery).

https://github.com/seglo/alpakka/commit/8cff3db2465a1abeb19fcbe45c64912abd859334

```
sbt:alpakka> google-cloud-bigquery/testOnly *.BigQueryQueriesSpec -- -z "should fail when failed future for comprehension"
[info] BigQueryQueriesSpec:
14:18:24.150 [pool-1-thread-1] INFO io.specto.hoverfly.junit.core.TempFileManager - Selecting the following binary based on the current operating system: hoverfly_linux_amd64
14:18:24.152 [pool-1-thread-1] INFO io.specto.hoverfly.junit.core.TempFileManager - Storing binary in temporary directory /tmp/hoverfly.2557597409281694884/hoverfly_linux_amd64
14:18:24.230 [pool-1-thread-1] INFO io.specto.hoverfly.junit.core.Hoverfly - Executing binary at /tmp/hoverfly.2557597409281694884/hoverfly_linux_amd64
14:18:24.234 [pool-1-thread-1] DEBUG org.zeroturnaround.exec.ProcessExecutor - Executing [/tmp/hoverfly.2557597409281694884/hoverfly_linux_amd64, -pp, 8500, -ap, 8888, -logs, json] in /tmp/hoverfly.2557597409281694884.
14:18:24.238 [pool-1-thread-1] DEBUG org.zeroturnaround.exec.ProcessExecutor - Started Process[pid=56875, exitValue="not exited"]
14:18:24.248 [Thread-3] INFO hoverfly - Default proxy port has been overwritten port=8500
14:18:24.250 [Thread-3] INFO hoverfly - Default admin port has been overwritten port=8888
14:18:24.250 [Thread-3] INFO hoverfly - Using memory backend
14:18:24.250 [Thread-3] INFO hoverfly - Proxy prepared... Destination=. Mode=simulate ProxyPort=8500
14:18:24.251 [Thread-3] INFO hoverfly - current proxy configuration destination=. mode=simulate port=8500
14:18:24.251 [Thread-3] INFO hoverfly - Admin interface is starting... AdminPort=8888
14:18:24.251 [Thread-3] INFO hoverfly - serving proxy
14:18:24.257 [Thread-3] INFO hoverfly - Mode has been changed mode=simulate
14:18:24.262 [pool-1-thread-1] INFO io.specto.hoverfly.junit.core.ProxyConfigurer - Setting proxy host to localhost
14:18:24.262 [pool-1-thread-1] INFO io.specto.hoverfly.junit.core.ProxyConfigurer - Setting proxy proxyPort to 8500
[info] BigQueryQueries
[info]   should get query results
[info]   should fail
[info]   - when failed future for comprehension *** FAILED *** (5 seconds, 53 milliseconds)
[info]     java.util.concurrent.TimeoutException: Futures timed out after [5 seconds]
[info]     at scala.concurrent.impl.Promise$DefaultPromise.ready(Promise.scala:259)
[info]     at scala.concurrent.impl.Promise$DefaultPromise.result(Promise.scala:263)
[info]     at scala.concurrent.Await$.$anonfun$result$1(package.scala:220)
[info]     at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:57)
[info]     at scala.concurrent.Await$.result(package.scala:146)
[info]     at akka.stream.alpakka.googlecloud.bigquery.scaladsl.BigQueryQueriesSpec.$anonfun$new$13(BigQueryQueriesSpec.scala:216)
[info]     at org.scalatest.wordspec.AsyncWordSpecLike.transformToOutcomeParam$1(AsyncWordSpecLike.scala:134)
[info]     at org.scalatest.wordspec.AsyncWordSpecLike.$anonfun$registerTestToRun$1(AsyncWordSpecLike.scala:135)
[info]     at org.scalatest.AsyncTestSuite.$anonfun$transformToOutcome$1(AsyncTestSuite.scala:240)
[info]     at org.scalatest.wordspec.AsyncWordSpecLike$$anon$3.apply(AsyncWordSpecLike.scala:1061)
[info]     at org.scalatest.AsyncTestSuite.withFixture(AsyncTestSuite.scala:313)
[info]     at org.scalatest.AsyncTestSuite.withFixture$(AsyncTestSuite.scala:312)
[info]     at akka.stream.alpakka.googlecloud.bigquery.scaladsl.BigQueryQueriesSpec.withFixture(BigQueryQueriesSpec.scala:33)
[info]     at org.scalatest.wordspec.AsyncWordSpecLike.invokeWithAsyncFixture$1(AsyncWordSpecLike.scala:1059)
[info]     at org.scalatest.wordspec.AsyncWordSpecLike.$anonfun$runTest$1(AsyncWordSpecLike.scala:1073)
[info]     at org.scalatest.AsyncSuperEngine.runTestImpl(AsyncEngine.scala:374)
[info]     at org.scalatest.wordspec.AsyncWordSpecLike.runTest(AsyncWordSpecLike.scala:1073)
[info]     at org.scalatest.wordspec.AsyncWordSpecLike.runTest$(AsyncWordSpecLike.scala:1053)
[info]     at akka.stream.alpakka.googlecloud.bigquery.scaladsl.BigQueryQueriesSpec.runTest(BigQueryQueriesSpec.scala:33)
[info]     at org.scalatest.wordspec.AsyncWordSpecLike.$anonfun$runTests$1(AsyncWordSpecLike.scala:1132)
[info]     at org.scalatest.AsyncSuperEngine.$anonfun$runTestsInBranch$1(AsyncEngine.scala:432)
[info]     at scala.collection.LinearSeqOptimized.foldLeft(LinearSeqOptimized.scala:126)
[info]     at scala.collection.LinearSeqOptimized.foldLeft$(LinearSeqOptimized.scala:122)
[info]     at scala.collection.immutable.List.foldLeft(List.scala:89)
[info]     at org.scalatest.AsyncSuperEngine.traverseSubNodes$1(AsyncEngine.scala:406)
[info]     at org.scalatest.AsyncSuperEngine.runTestsInBranch(AsyncEngine.scala:479)
[info]     at org.scalatest.AsyncSuperEngine.$anonfun$runTestsInBranch$1(AsyncEngine.scala:460)
[info]     at scala.collection.LinearSeqOptimized.foldLeft(LinearSeqOptimized.scala:126)
[info]     at scala.collection.LinearSeqOptimized.foldLeft$(LinearSeqOptimized.scala:122)
[info]     at scala.collection.immutable.List.foldLeft(List.scala:89)
[info]     at org.scalatest.AsyncSuperEngine.traverseSubNodes$1(AsyncEngine.scala:406)
[info]     at org.scalatest.AsyncSuperEngine.runTestsInBranch(AsyncEngine.scala:479)
[info]     at org.scalatest.AsyncSuperEngine.$anonfun$runTestsInBranch$1(AsyncEngine.scala:460)
[info]     at scala.collection.LinearSeqOptimized.foldLeft(LinearSeqOptimized.scala:126)
[info]     at scala.collection.LinearSeqOptimized.foldLeft$(LinearSeqOptimized.scala:122)
[info]     at scala.collection.immutable.List.foldLeft(List.scala:89)
[info]     at org.scalatest.AsyncSuperEngine.traverseSubNodes$1(AsyncEngine.scala:406)
[info]     at org.scalatest.AsyncSuperEngine.runTestsInBranch(AsyncEngine.scala:487)
[info]     at org.scalatest.AsyncSuperEngine.runTestsImpl(AsyncEngine.scala:555)
[info]     at org.scalatest.wordspec.AsyncWordSpecLike.runTests(AsyncWordSpecLike.scala:1132)
[info]     at org.scalatest.wordspec.AsyncWordSpecLike.runTests$(AsyncWordSpecLike.scala:1131)
[info]     at akka.stream.alpakka.googlecloud.bigquery.scaladsl.BigQueryQueriesSpec.runTests(BigQueryQueriesSpec.scala:33)
[info]     at org.scalatest.Suite.run(Suite.scala:1112)
[info]     at org.scalatest.Suite.run$(Suite.scala:1094)
[info]     at akka.stream.alpakka.googlecloud.bigquery.scaladsl.BigQueryQueriesSpec.org$scalatest$wordspec$AsyncWordSpecLike$$super$run(BigQueryQueriesSpec.scala:33)
[info]     at org.scalatest.wordspec.AsyncWordSpecLike.$anonfun$run$1(AsyncWordSpecLike.scala:1177)
[info]     at org.scalatest.AsyncSuperEngine.runImpl(AsyncEngine.scala:625)
[info]     at org.scalatest.wordspec.AsyncWordSpecLike.run(AsyncWordSpecLike.scala:1177)
[info]     at org.scalatest.wordspec.AsyncWordSpecLike.run$(AsyncWordSpecLike.scala:1175)
[info]     at akka.stream.alpakka.googlecloud.bigquery.scaladsl.BigQueryQueriesSpec.org$scalatest$BeforeAndAfterAll$$super$run(BigQueryQueriesSpec.scala:33)
[info]     at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:213)
[info]     at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
[info]     at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
[info]     at akka.stream.alpakka.googlecloud.bigquery.scaladsl.BigQueryQueriesSpec.run(BigQueryQueriesSpec.scala:33)
[info]     at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:318)
[info]     at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:513)
[info]     at sbt.ForkMain$Run.lambda$runTest$1(ForkMain.java:413)
[info]     at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[info]     at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[info]     at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[info]     at java.base/java.lang.Thread.run(Thread.java:834)
14:18:29.384 [pool-1-thread-1] INFO io.specto.hoverfly.junit.core.Hoverfly - Destroying hoverfly process
14:18:29.385 [WaitForProcess-Process[pid=56875, exitValue="not exited"]] DEBUG org.zeroturnaround.exec.WaitForProcess - Process[pid=56875, exitValue=143] stopped with exit code 143
[info] ScalaTest
[info] Run completed in 6 seconds, 473 milliseconds.
[info] Total number of tests run: 1
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 0, failed 1, canceled 0, ignored 0, pending 0
[info] *** 1 TEST FAILED ***
[error] Failed: Total 1, Failed 1, Errors 0, Passed 0
[error] Failed tests:
[error]         akka.stream.alpakka.googlecloud.bigquery.scaladsl.BigQueryQueriesSpec
[error] (google-cloud-bigquery / Test / testOnly) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 7 s, completed Feb. 23, 2021, 2:18:29 p.m.
sbt:alpakka>
```

Reply:
Ok thanks, when I run your test it times out as you say. Still confused why this doesn't happen when I run your snippet in a new project or why it happens at all. I'm going to poke around with this a bit.

Reply:
Ok, I can get your test the work by adding `import system.dispatcher`. The timeout you observe seems to occur when using the `SerialExecutionContext` provided by `AsyncTestSuite`.

Reply:
Got it. Thanks for the explanation.